alf/algorithms/doac_algorithm.py

```diff
@@ -0,0 +1,263 @@
+            name=name)
+
+        assert self._act_type == ActionType.Continuous, (
+            "Only continuous action space is supported for DOacAlgorithm.")
+        self._explore_delta = explore_delta
+        self._beta_ub = beta_ub
+
+        self._explore_tau_hat, _ = self._get_tau(
+            batch_size=1, tau_type='fixed')
+
+    def _get_q_ub_from_critics(self, observation, action, state):
+        critic_inputs = (observation, action)
+        critics_quantiles, critic_state = self._critic_networks(
+            (critic_inputs, self._explore_tau_hat), state=state)
+        critics = critics_quantiles.mean(-1)
+        assert critics.ndim == 2
+        q_mean = critics.mean(dim=1)
+        q_std = critics.std(1, unbiased=True)
+        q_ub = q_mean + self._beta_ub * q_std
+
+        return q_ub, critic_state
+
+    def _predict_action(self,
+                        observation,
+                        state: SacActionState,
+                        epsilon_greedy=None,
+                        eps_greedy_sampling=False,
+                        rollout=False):
+        """
+        Differences between DSacAlgorithm._predict_action:
+
+        1. Only continuous actions are supported.
+
+        2. OAC explore policy is constructed from the target policy (actor_network) 
+        and used for action prediction during rollout.
+        """
+
+        # new_state = SacActionState()
+        # action_dist, actor_network_state = self._actor_network(
+        #     observation, state=state.actor_network)
+        # assert isinstance(action_dist, td.TransformedDistribution), (
+        #     "Squashed distribution is expected from actor_network.")
+        # assert isinstance(
+        #     action_dist.base_dist, dist_utils.DiagMultivariateNormal
+        # ), ("the base distribution should be diagonal multivariate normal.")
+        # normal_dist = action_dist.base_dist
+        # unsquashed_mean = normal_dist.mean
+        # unsquashed_std = normal_dist.stddev
+        # unsquashed_var = normal_dist.variance
+        # new_state = new_state._replace(actor_network=actor_network_state)
+
+        new_state = SacActionState()
+        action_dist, actor_network_state = self._actor_network(
+            observation, state=state.actor_network)
+        new_state = new_state._replace(actor_network=actor_network_state)
+
+        action_dist_type = nest.map_structure(
+            lambda dist: isinstance(dist, td.TransformedDistribution),
+            action_dist)
+        action_base_dist_type = nest.map_structure(
+            lambda dist: isinstance(dist.base_dist, dist_utils.
+                                    DiagMultivariateNormal), action_dist)
+        action_dist_type = nest.flatten(action_dist_type)
+        action_base_dist_type = nest.flatten(action_base_dist_type)
+        assert all(nest.flatten(action_dist_type)), (
+            "Squashed distribution is expected from actor_network.")
```

Comment:
If line 181-182 is to ensure the distributions are squashed, the current code block does not seem to be able to ensure this? Currently it only checks if the base_dist is DiagMultiNormal, that does not have any guarantee on the outer distribution containing the base_dist?

Reply:
The ``action_dist_type`` extracted from line 173-175 was checked for ``TransformedDistribution``?

Reply:
You are right.